crates/starfish/core/src/block_manager/mod.rs

```diff
@@ -135,26 +139,56 @@ impl BlockManager {
+    fn write_block_headers_and_transactions_to_dag_state(
+        &self,
+        block_headers: Vec<VerifiedBlockHeader>,
+        transactions: Vec<VerifiedTransactions>,
+    ) {
+        let mut write_guard = self.dag_state.write();
+        write_guard.accept_block_headers(block_headers);
+        for verified_transaction in transactions {
+            write_guard.add_transactions(verified_transaction, "Block streaming");
+        }
+    }
 
-        // Insert the accepted blocks into DAG state so future blocks including them as
-        // ancestors do not get suspended.
-        self.dag_state
-            .write()
-            .accept_block_headers(accepted_block_headers.clone());
-
-        // check if we already have blocks for this accepted header. If yes, add them to
-        // dag_state
-        for block_header in accepted_block_headers.iter() {
-            if let Some(block) = self.suspended_blocks.remove(&block_header.reference()) {
-                // for this accepted header we already have a block, so we add it to dag_state
-                self.dag_state
-                    .write()
-                    .add_transactions(block.verified_transactions, "Block streaming");
+    /// Resolves transactions from the provided blocks and accepted block
+    /// headers.
+    ///
+    /// Moves transactions from suspended blocks whose headers are now accepted,
+    /// and optionally processes newly received blocks, adding their
+    /// transactions if accepted or re-suspending them otherwise.
+    fn resolve_transactions(
+        &mut self,
+        block_headers_to_be_accepted: &[VerifiedBlockHeader],
+        blocks: Option<Vec<VerifiedBlock>>,
+    ) -> Vec<VerifiedTransactions> {
+        let block_refs_to_be_accepted = block_headers_to_be_accepted
+            .iter()
+            .map(|h| h.reference())
+            .collect::<BTreeSet<_>>();
+        let mut accepted_transactions_from_suspended_blocks = vec![];
+        for block_ref in block_refs_to_be_accepted.iter() {
+            if let Some(block) = self.suspended_blocks.remove(block_ref) {
+                // for this accepted header we already have a block, so we add it accepted
+                // transactions
+                accepted_transactions_from_suspended_blocks.push(block.verified_transactions);
             }
         }
 
-        // Figure out the new missing blocks
-        (accepted_block_headers, ancestors_to_fetch)
+        if let Some(blocks) = blocks {
+            let mut accepted_transactions_from_blocks = vec![];
+            for block in blocks {
+                if block_refs_to_be_accepted.contains(&block.reference()) {
+                    accepted_transactions_from_blocks.push(block.verified_transactions);
+                } else {
+                    self.suspended_blocks.insert(block.reference(), block);
+                }
+            }
+            accepted_transactions_from_suspended_blocks.extend(accepted_transactions_from_blocks);
```

Comment:
`accepted_transactions_from_suspended_blocks` this name is a bit confusing if also used here with non-suspended blocks. Perhaps rename to `accepted_transactions` or `all_accepted_transactions`